controllers/ingressnodefirewallconfig_controller.go

```diff
@@ -0,0 +1,137 @@
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the IngressNodeFirewallConfig object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.12.1/pkg/reconcile
+func (r *IngressNodeFirewallConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = context.Background()
+	logger := r.Log.WithValues("ingress node firewall config", req.NamespacedName)
+
+	instance := &ingressnodefwv1alpha1.IngressNodeFirewallConfig{}
+	err := r.Get(context.TODO(), req.NamespacedName, instance)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	if req.Name != defaultIngressNodeFirwallCrName {
+		logger.Error(err, "Invalid IngressNode firewall config resource name", "name", req.Name)
+		return ctrl.Result{}, nil // Return success to avoid requeue
+	}
+
+	if err := r.syncIngressNodeFwConfigResources(instance); err != nil {
+		return ctrl.Result{}, nil
+	}
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *IngressNodeFirewallConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&ingressnodefwv1alpha1.IngressNodeFirewallConfig{}).
+		Complete(r)
+}
+
+func (r *IngressNodeFirewallConfigReconciler) syncIngressNodeFwConfigResources(config *ingressnodefwv1alpha1.IngressNodeFirewallConfig) error {
+	logger := r.Log.WithName("syncIngressNodeFirewallConfigResources")
+	logger.Info("Start")
+	data := render.MakeRenderData()
+
+	data.Data["Image"] = os.Getenv("DAEMONSET_IMAGE")
+	objs, err := render.RenderDir(ManifestPath, &data)
+	if err != nil {
+		logger.Error(err, "Fail to render config daemon manifests")
+		return err
+	}
+
+	for _, obj := range objs {
+		if obj.GetKind() == "DaemonSet" &&
+			(len(config.Spec.NodeSelector) > 0 || len(config.Spec.Tolerations) > 0) {
+			scheme := kscheme.Scheme
+			ds := &appsv1.DaemonSet{}
+			err = scheme.Convert(obj, ds, nil)
+			if err != nil {
+				logger.Error(err, "Fail to convert IngressNodeFirewallConfig object to DaemonSet")
+				return err
+			}
+			if len(config.Spec.NodeSelector) > 0 {
+				ds.Spec.Template.Spec.NodeSelector = config.Spec.NodeSelector
+			}
+			if len(config.Spec.Tolerations) > 0 {
+				ds.Spec.Template.Spec.Tolerations = config.Spec.Tolerations
+			}
+			err = scheme.Convert(ds, obj, nil)
```

Comment:
Right above this line, add:
~~~
ctrl.SetControllerReference(config, ds, r.Scheme)
~~~

This will set the owner reference to the DS and it will allow us to monitor that DS's updates, deletes, etc:
~~~
    ownerReferences:
    - apiVersion: ingress-nodefw.ingress-nodefw/v1alpha1
      blockOwnerDeletion: true
      controller: true
      kind: IngressNodeFirewallConfig
      name: ingressnodefirewallconfig
      uid: b5c6875e-7ae0-4daa-8cea-fae191ff14f7
~~~

Reply:
And then with those 2 changes (Owns and the owner reference), the DS is now immediately recreated or updated when something happens to it:
~~~
[akaris@linux ingress-node-firewall (daemon)]$ oc get ds
NAME                           DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR            AGE
ingress-node-firewall-daemon   3         3         3       3            3           kubernetes.io/os=linux   7m25s
[akaris@linux ingress-node-firewall (daemon)]$ oc delete ds ingress-node-firewall-daemon
daemonset.apps "ingress-node-firewall-daemon" deleted
[akaris@linux ingress-node-firewall (daemon)]$ oc get ds
NAME                           DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR            AGE
ingress-node-firewall-daemon   3         3         2       3            2           kubernetes.io/os=linux   4s
[akaris@linux ingress-node-firewall (daemon)]$ oc get ds
NAME                           DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR            AGE
ingress-node-firewall-daemon   3         3         3       3            3           kubernetes.io/os=linux   7s
~~~